docs/architecture/adr-008-event-subscription.md

```diff
@@ -0,0 +1,178 @@
+
+The entities in the diagram below are described in the following subsections.
+The diagram also gives some indication as to the proposed concurrency model for
+the architecture and how the entities relate to each other.
+
+![](assets/rpc-client-subscription-rels.png)
+
+### Subscription
+
+A `Subscription` here is envisaged as an entity that implements the
+[Stream][futures-stream] trait, allowing its owner to asynchronously iterate
+through all of the **relevant** incoming events. Use of such a subscription
+should be as simple as:
+
+```rust
+while let Some(result_event) = subscription.next().await {
+   match result_event {
+      Ok(event) => { /* handle event */ },
+      Err(e) => { /* terminate subscription and report error */ },
+   }
+}
+```
+
+Since the underlying events may not be produced at the same cadence as that of
+the `Subscription` owner's processing, it is proposed here to make use of an
+appropriate type of buffered channel to realize this. The size of this buffer
+must be configurable through the RPC Client.
+
+### RPC Pub/Sub
+
+The RPC Client must asynchronously spawn a component `PubSub` whose
+responsibilities are:
+
+1. Manage subscriptions (subscribe/unsubscribe operations from the RPC Client).
+2. Receive and deserialize incoming `Event`s.
+3. Route `Event`s to `Subscription`s whose `Query` matches the `Event`.
+4. Manage transport-level concerns (i.e. WebSockets connection: keep-alive
+   messages, graceful connection closure, etc.).
+
+Items (3) and (4) could be encapsulated into their own sub-components to allow
+for better separation of concerns and therefore easier testing. For example, (3)
+could be encapsulated into a `PubSubRouter` component and (4) into a component
+that only handles transport-level concerns.
+
+This is an *internal* component whose interface only matters to the RPC Client.
+
+### RPC Client Interface Extension
+
+The following interface is proposed as an extension to the existing [`Client`
+structure][client]:
+
+```rust
+impl Client {
+    /// Initiates a subscription for events produced by the given query. More
+    /// details as to the query syntax is available from the [Tendermint
+    /// `/subscribe` RPC endpoint documentation][tm-rpc-subscribe].
+    /// 
+    /// [tm-rpc-subscribe]: https://docs.tendermint.com/master/rpc/#/Websocket/subscribe
+    pub async fn subscribe(&mut self, query: &Query) -> Result<Subscription, Error> {
+        // ...
+    }
+
+    /// Terminates the given subscription and consumes it.
+    pub async fn unsubscribe(&mut self, subs: Subscription) -> Result<(), Error> {
+       // ...
+    }
+}
+```
+
+### Query
+
+Due to the potential complexity of queries, the proposed `Query` entity should
+functionally take its lead from the [Go-based `Query` interface][tm-go-query].
+This type of interface allows for simple short-term implementation (e.g. just
+querying by way of event type), with eventual full implementation of the [query
+PEG][query-peg].
```

Comment:
If this PEG "source-of-truth" exists, why aren't we going to aim for support of it from the get-go?

Reply:
I thought we could do that in a separate ADR? I'll be expanding on this ADR a little more shortly, so I thought I wouldn't muddy the waters with too much focus on the querying.

Reply:
Super fair, let's capture a follow-up.